# Performing text analysis on recipe ingredients + cuisines

**Topics:** Text analysis, classifiers

**Datasets**

- **recipes.csv** Once upon a time I scraped a bunch of recipes, not sure where they're from.

## Imports

We're going to be using

* Pandas
* Regular expressions
* Decision Trees
* Random Forests
* Logistic Regression
* Confusion matrices
* CountVectorizer
* TfidfVectorizer

Import everything you need for those. It's easier than doing them one-by-one later, I think! We're also going to be doing the graphviz tree visualization, but I always just cut and paste that entire block as a unit.

In [1]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import re
from nltk.stem.porter import PorterStemmer
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier


%matplotlib inline

## Read in our data

There's nothing too complicated about it.

In [2]:
df = pd.read_csv('recipes.csv')
df.head()

cuisine     id                                    ingredient_list
0        greek  10259  romaine lettuce, black olives, grape tomatoes,...
1  southern_us  25693  plain flour, ground pepper, salt, tomatoes, gr...
2     filipino  20130  eggs, pepper, salt, mayonaise, cooking oil, gr...
3       indian  22213                  water, vegetable oil, wheat, salt
4       indian  13162  black pepper, shallots, cornflour, cayenne pep...

# Vectorizing the ingredient list

These words aren't going to do us much good unless we can convert them to text. Use the `CountVectorizer` to count the number of times each word appears in each sentence.

In [3]:
count_vectorizer = CountVectorizer()

In [4]:
X = count_vectorizer.fit_transform(df.ingredient_list)

In [5]:
X

<39774x3010 sparse matrix of type '<class 'numpy.int64'>'
	with 761951 stored elements in Compressed Sparse Row format>

In [6]:
words_df = pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names())

In [7]:
words_df.head()

00  10  100  14  15  25  33  40  43  95  ...  za  zatarain  zatarains  \
0   0   0    0   0   0   0   0   0   0   0  ...   0         0          0   
1   0   0    0   0   0   0   0   0   0   0  ...   0         0          0   
2   0   0    0   0   0   0   0   0   0   0  ...   0         0          0   
3   0   0    0   0   0   0   0   0   0   0  ...   0         0          0   
4   0   0    0   0   0   0   0   0   0   0  ...   0         0          0   

   zero  zest  zesty  zinfandel  ziti  zucchini  épices  
0     0     0      0          0     0         0       0  
1     0     0      0          0     0         0       0  
2     0     0      0          0     0         0       0  
3     0     0      0          0     0         0       0  
4     0     0      0          0     0         0       0  

[5 rows x 3010 columns]

Your result should be a dataframe named `words_df` that looks like a much wider version of this:
    
|00|10|...|zucchini|épices|
|---|---|---|---|---|
|0|0|...|0|0|
|0|0|...|0|0|
|0|0|...|0|0|
|0|0|...|0|0|
|0|0|...|0|0|

## Counting

### Count up how many times each word is used

Don't think too hard about this, there's nothing fancy about how you do this.

* **Tip:** Just imagine it's a normal dataframe and you want to add up each column.
* **Tip:** ...or imagine you're trying to add up _one_ column but you forget the column name.

In [8]:
words_df.sum()

00                  1
10                  5
100                 1
14                  5
15                  3
25                  3
33                  6
40                 12
43                  1
95                  4
96                  1
abalone             2
abbamele            1
absinthe            1
abura               4
acai                5
accent              8
accompaniment       1
achiote            23
acid                6
acini               3
ackee               9
acorn              18
acting             21
activ               1
active            379
added              49
adobo             314
adzuki              5
agar                8
                 ... 
yams               54
yardlong           22
yeast             758
yellow           2299
yellowfin           4
yellowtail          2
yoghurt           279
yogurt           1167
yolk               14
yolks            1117
yoplait             3
york               20
young               4
yu                  3
yuca      

### What are the top 10 most common ingredients?

You'll want to use the results from the last question. Again, don't overthink it.

In [9]:
words_df.sum().sort_values(ascending=False).head(10)

pepper    25742
salt      24462
oil       23344
garlic    18941
ground    18271
fresh     17853
sauce     13129
sugar     12493
onions    12341
cheese    11776
dtype: int64

### Does anything seem wrong with those ingredients?

Do they all seem very ingredient-y to you?

### Look at your original recipes dataframe again

What do the ingredients look like? How would you as a _person_ separate them compared to how the vectorizer did it?

# Using a custom tokenizer

Usually we can trust what the vectorizer does to the words, but this time we can't! Since we have a better idea about how to split it, we use a custom function to split each piece of text up.

To do this: copy and paste your code where you counted words up above. Add `tokenizer=split_on_commas` to `CountVectorizer` to tell it to use my custom function below.

In [10]:
def split_on_commas(text):
    # Split the list on commas, then remove any extra whitespace
    pieces = text.split(",")
    no_spaces = [piece.strip() for piece in pieces if piece]
    return no_spaces

count_vectorizer = CountVectorizer(tokenizer=split_on_commas)
X = count_vectorizer.fit_transform(df.ingredient_list)
print(count_vectorizer.get_feature_names()) 

['(    oz.) tomato sauce', '(   oz.) tomato paste', '(10 oz.) frozen chopped spinach', '(14 oz.) sweetened condensed milk', '(14.5 oz.) diced tomatoes', '(15 oz.) refried beans', '1 inch thick', '1% low-fat buttermilk', '1% low-fat chocolate milk', '1% low-fat cottage cheese', '1% low-fat milk', '2 1/2 to 3 lb. chicken', '2% low fat cheddar chees', '2% low-fat cottage cheese', '2% lowfat greek yogurt', '2% milk shredded mozzarella cheese', '2% reduced-fat milk', '25% less sodium chicken broth', '33% less sodium cooked deli ham', '33% less sodium cooked ham', '33% less sodium ham', '33% less sodium smoked fully cooked ham', '40% less sodium taco seasoning', '40% less sodium taco seasoning mix', '7 up', '8 ounc ziti pasta', '95% lean ground beef', 'a taste of thai rice noodles', 'abalone', 'abbamele', 'absinthe', 'abura age', 'acai juice', 'accent', 'accent seasoning', 'accompaniment', 'achiote', 'achiote paste', 'achiote powder', 'acini di pepe', 'ackee', 'acorn squash', 'active dry yea

Doesn't that look nicer??

_Yes!_

## Word counting, again

Get the top 15 most common ingredients. **It should be a much nicer list!**

In [11]:
words2 = pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names())
words2.head()

(    oz.) tomato sauce  (   oz.) tomato paste  \
0                       0                      0   
1                       0                      0   
2                       0                      0   
3                       0                      0   
4                       0                      0   

   (10 oz.) frozen chopped spinach  (14 oz.) sweetened condensed milk  \
0                                0                                  0   
1                                0                                  0   
2                                0                                  0   
3                                0                                  0   
4                                0                                  0   

   (14.5 oz.) diced tomatoes  (15 oz.) refried beans  1 inch thick  \
0                          0                       0             0   
1                          0                       0             0   
2                          0                       0             0   
3                          0                       0             0   
4                          0                       0             0   

   1% low-fat buttermilk  1% low-fat chocolate milk  \
0                      0                          0   
1                      0                          0   
2                      0                          0   
3                      0                          0   
4                      0                          0   

   1% low-fat cottage cheese  ...  yuzukosho  za'atar  \
0                          0  ...          0        0   
1                          0  ...          0        0   
2                          0  ...          0        0   
3                          0  ...          0        0   
4                          0  ...          0        0   

   zatarains creole seasoning  zatarain’s jambalaya mix  zest  \
0                           0                         0     0   
1                           0                         0     0   
2                           0                         0     0   
3                           0                         0     0   
4                           0                         0     0   

   zesty italian dressing  zinfandel  ziti  zucchini  zucchini blossoms  
0                       0          0     0         0                  0  
1                       0          0     0         0                  0  
2                       0          0     0         0                  0  
3                       0          0     0         0                  0  
4                       0          0     0         0                  0  

[5 rows x 6711 columns]

In [12]:
words2.sum().sort_values(ascending=False).head(15)

salt                   18049
olive oil               7972
onions                  7972
water                   7457
garlic                  7380
sugar                   6434
garlic cloves           6237
butter                  4848
ground black pepper     4785
all-purpose flour       4632
pepper                  4438
vegetable oil           4385
eggs                    3388
soy sauce               3296
kosher salt             3113
dtype: int64

# Using our cuisines

## What are the most popular cuisines?

* **Tip:** You'll need to go back to your original dataframe for this.

In [13]:
df.cuisine.value_counts()

italian         7838
mexican         6438
southern_us     4320
indian          3003
chinese         2673
french          2646
cajun_creole    1546
thai            1539
japanese        1423
greek           1175
spanish          989
korean           830
vietnamese       825
moroccan         821
british          804
filipino         755
irish            667
jamaican         526
russian          489
brazilian        467
Name: cuisine, dtype: int64

## Find every recipe that uses soy sauce

In [14]:
df[df.ingredient_list.str.contains('soy sauce')].groupby(by='cuisine').size()

cuisine
brazilian          3
british            3
cajun_creole      13
chinese         2028
filipino         284
french             6
greek              6
indian            24
irish              7
italian           21
jamaican          80
japanese         755
korean           531
mexican           54
moroccan           4
russian            3
southern_us       29
spanish            3
thai             489
vietnamese       227
dtype: int64

## What's are the top 5 cuisines that uses soy sauce?

In [15]:
df[df.ingredient_list.str.contains('soy sauce')].groupby(by='cuisine').size().sort_values(ascending=False).head(5)

cuisine
chinese     2028
japanese     755
korean       531
thai         489
filipino     284
dtype: int64

Hm, soy sauce seems pretty common in Chinese food. **Maybe we could do something with that?**

# Building a Chinese food detector

Create a new column in your original dataframe called `is_chinese`. It should be `1` if the recipe is Chinese, `0` if the recipe is not.

In [16]:
import numpy as np
df['is_chinese'] = np.where(df['cuisine'] == 'chinese', 1,0)

Confirm that you have 2,673 Chinese recipes and 37,101 non-Chinese recipes.

In [17]:
df.is_chinese.value_counts()

0    37101
1     2673
Name: is_chinese, dtype: int64

## Building your features

To use a classifier, we need to build an `X` and a `y`.

* Your `X` is going to be the word counts for each recipe
* Your `y` is going to be whether each recipe Chinese or not

This is going to be a little different than what we normally do. Typically we use `.drop(column=` to remove our label, but this time we have **two separate dataframes** we're managing, so don't be surprised if you don't sue it.

In [18]:
X = words2
y = df['is_chinese']

Confirm that `X` is a dataframe of numbers (mostly zeroes) and `y` is a single column of `0`s and `1`s.

In [19]:
X.head()

(    oz.) tomato sauce  (   oz.) tomato paste  \
0                       0                      0   
1                       0                      0   
2                       0                      0   
3                       0                      0   
4                       0                      0   

   (10 oz.) frozen chopped spinach  (14 oz.) sweetened condensed milk  \
0                                0                                  0   
1                                0                                  0   
2                                0                                  0   
3                                0                                  0   
4                                0                                  0   

   (14.5 oz.) diced tomatoes  (15 oz.) refried beans  1 inch thick  \
0                          0                       0             0   
1                          0                       0             0   
2                          0                       0             0   
3                          0                       0             0   
4                          0                       0             0   

   1% low-fat buttermilk  1% low-fat chocolate milk  \
0                      0                          0   
1                      0                          0   
2                      0                          0   
3                      0                          0   
4                      0                          0   

   1% low-fat cottage cheese  ...  yuzukosho  za'atar  \
0                          0  ...          0        0   
1                          0  ...          0        0   
2                          0  ...          0        0   
3                          0  ...          0        0   
4                          0  ...          0        0   

   zatarains creole seasoning  zatarain’s jambalaya mix  zest  \
0                           0                         0     0   
1                           0                         0     0   
2                           0                         0     0   
3                           0                         0     0   
4                           0                         0     0   

   zesty italian dressing  zinfandel  ziti  zucchini  zucchini blossoms  
0                       0          0     0         0                  0  
1                       0          0     0         0                  0  
2                       0          0     0         0                  0  
3                       0          0     0         0                  0  
4                       0          0     0         0                  0  

[5 rows x 6711 columns]

In [20]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: is_chinese, dtype: int64

### Split into test and train datasets

If we train our classifier on all of our recipes, it might cheat and memorize the answers (aka overfitting). Instead, we'll need to use `train_test_split` to separate into a training set it can study, and a test set we can use to see how well it performs. 

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y)

# Using our classifiers

## Decision Tree

**Create a decision tree classifier, and train it on our training data.** You can make a decision tree classifier like this:

```python
clf = DecisionTreeClassifier()
```

You'll want to also give it `max_depth=5` so we can visualize it without too much trouble later - otherwise it'll make many many decisions which will make a large complicated graphic.

Additionally, put `%%time` in the first row of the cell to see how long it takes to train our classifier.

In [22]:
%%time
clf = tree.DecisionTreeClassifier(max_depth=5)
clf = clf.fit(X_train, y_train)

CPU times: user 30.5 s, sys: 2.3 s, total: 32.8 s
Wall time: 35.8 s


In [23]:
clf

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=5,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

### Print out our confusion matrix 

Use the one from our notes and homework, not the default scikit-learn version. The output should look something like this:

|.|Predicted not chinese|Predicted chinese|
|---|---|---|
|Is not chinese|9209|60|
|Is chinese|377|298|

In [24]:
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not chinese', 'chinese'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted not chinese  Predicted chinese
Is not chinese                   9197                 72
Is chinese                        378                297

**How do we feel about our classifier's performance?** Describe what is happening in the "is chinese" row using words.

In [25]:
# it's bad at predicting what is chinese, good at predicting what's not

### What are the important features?

You can use the following code to see feature importance for a decision tree.

```python
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)
```

In [26]:
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)

feature  feature importance
5667                     soy sauce            0.409496
3535               light soy sauce            0.131039
1552                   corn starch            0.125814
3658          low sodium soy sauce            0.094856
1188     chinese five-spice powder            0.059192
5357                    sesame oil            0.043819
5373                 shaoxing wine            0.034400
2986                  hoisin sauce            0.030889
3978                         mirin            0.020082
4281                  oyster sauce            0.018142
2231                    fish sauce            0.017040
1124                 chicken wings            0.004122
954                        carrots            0.002945
5216                          salt            0.001579
5749                      sriracha            0.001508
1799                dark soy sauce            0.001242
628                  bonito flakes            0.001186
2794           ground black pepper            0.001160
1086               chicken carcass            0.000828
6419                         water            0.000663
4502                      piecrust            0.000000
4503                       pierogi            0.000000
4480                 pickle wedges            0.000000
4479                 pickle spears            0.000000
4478                 pickle relish            0.000000
4477                  pickle juice            0.000000
4504                          pies            0.000000
4497                 pie crust mix            0.000000
4476              pickapeppa sauce            0.000000
4475              picholine olives            0.000000
...                            ...                 ...
2211  finely chopped fresh parsley            0.000000
2210                 fine sea salt            0.000000
2209                     fine salt            0.000000
2232                   fish steaks            0.000000
2234                      fishcake            0.000000
2258                 flavored wine            0.000000
2247                      flat cut            0.000000
2257                flavored vodka            0.000000
2256              flavored vinegar            0.000000
2255       flavored tortilla chips            0.000000
2254                flavored syrup            0.000000
2253             flavored rice mix            0.000000
2252                  flavored oil            0.000000
2251                     flatbread            0.000000
2250             flat leaf spinach            0.000000
2249             flat leaf parsley            0.000000
2248              flat iron steaks            0.000000
2246                  flat anchovy            0.000000
2235           fisher pecan halves            0.000000
2245            flanken short ribs            0.000000
2244                   flank steak            0.000000
2243                          flan            0.000000
2242                   flaked oats            0.000000
2241                flaked coconut            0.000000
2240                        flaked            0.000000
2239                     flageolet            0.000000
2238             five-spice powder            0.000000
2237                    five spice            0.000000
2236                 fisher pecans            0.000000
6710             zucchini blossoms            0.000000

[6711 rows x 2 columns]

We gave it a lot of columns to pay attention to, so **why are so many of them zero?**

In [27]:
# we didn't specify a max_depth

### Visualize the decision tree

While those numbers are fun (if a little long), the _real fun_ is visualizing the tree itself. You can use this code to do that:

```python
from sklearn import tree
import graphviz

label_names = ['not chinese', 'chinese']
feature_names = X.columns

dot_data = tree.export_graphviz(clf,
                    feature_names=feature_names,  
                    filled=True,
                    class_names=label_names)  
graph = graphviz.Source(dot_data)  
graph
```

In [28]:
# sorry, graphviz hates me!

**What do you think about that?** If you know about cooking or ingredients, do you recognize what some of those splits mean in terms of different cuisines?

## Random Forest

A random forest is like making a bunch of decision trees work together. **Create a random forest classifier, and train it on our data.**

```python
clf = RandomForestClassifier()
```

Right now, the `RandomForestClassifier` uses 10 decision trees to make its decision, with `n_estimators=10` being the default. You'll see a warning about how it's going to change to `100` in a future version.

You can try to use `100` now, but notice how long it takes! Use `%%time` to time how long it takes to train the random forest.

In [29]:
model = RandomForestClassifier(n_estimators = 10, max_depth=5)
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=5, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

### What are the feature importances?

You can re-use your decision tree code. Are the results any different?

In [30]:
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)

feature  feature importance
5667                     soy sauce            0.409496
3535               light soy sauce            0.131039
1552                   corn starch            0.125814
3658          low sodium soy sauce            0.094856
1188     chinese five-spice powder            0.059192
5357                    sesame oil            0.043819
5373                 shaoxing wine            0.034400
2986                  hoisin sauce            0.030889
3978                         mirin            0.020082
4281                  oyster sauce            0.018142
2231                    fish sauce            0.017040
1124                 chicken wings            0.004122
954                        carrots            0.002945
5216                          salt            0.001579
5749                      sriracha            0.001508
1799                dark soy sauce            0.001242
628                  bonito flakes            0.001186
2794           ground black pepper            0.001160
1086               chicken carcass            0.000828
6419                         water            0.000663
4502                      piecrust            0.000000
4503                       pierogi            0.000000
4480                 pickle wedges            0.000000
4479                 pickle spears            0.000000
4478                 pickle relish            0.000000
4477                  pickle juice            0.000000
4504                          pies            0.000000
4497                 pie crust mix            0.000000
4476              pickapeppa sauce            0.000000
4475              picholine olives            0.000000
...                            ...                 ...
2211  finely chopped fresh parsley            0.000000
2210                 fine sea salt            0.000000
2209                     fine salt            0.000000
2232                   fish steaks            0.000000
2234                      fishcake            0.000000
2258                 flavored wine            0.000000
2247                      flat cut            0.000000
2257                flavored vodka            0.000000
2256              flavored vinegar            0.000000
2255       flavored tortilla chips            0.000000
2254                flavored syrup            0.000000
2253             flavored rice mix            0.000000
2252                  flavored oil            0.000000
2251                     flatbread            0.000000
2250             flat leaf spinach            0.000000
2249             flat leaf parsley            0.000000
2248              flat iron steaks            0.000000
2246                  flat anchovy            0.000000
2235           fisher pecan halves            0.000000
2245            flanken short ribs            0.000000
2244                   flank steak            0.000000
2243                          flan            0.000000
2242                   flaked oats            0.000000
2241                flaked coconut            0.000000
2240                        flaked            0.000000
2239                     flageolet            0.000000
2238             five-spice powder            0.000000
2237                    five spice            0.000000
2236                 fisher pecans            0.000000
6710             zucchini blossoms            0.000000

[6711 rows x 2 columns]

### Why can't we draw a tree for a random forest?

# Classifying with a logistic regression

To be a real data scientist, you throw every kind of classifier at your project until you find the one that works the best. Is that science? I don't know, really, but it's definitely reality.

## Building our classifier

This is the standard way to build a logistic regression classifier:

```python
clf = LogisticRegression(C=1e9, solver='lbfgs')
```

> You might get an error when fitting, though, that says "Increase the number of iterations." This means the regression failed to find a good answer, and you need to tell it to work harder. To do that, you need to set `max_iter=` to a higher number.
>
> The default is 100, and in previous homeworks we've cranked it up to 4000 with no problem. This time it isn't so easy - we have a lot lot lot lot of features this time, though, and every iteration makes the regression take more time until we get very very bored of things.
>
> You can solve this three ways:
>
> * Scroll up and add a `max_features=` to your vectorizer so it only picks a certain (lower) number of features, then re-run
> * Be very patient
> * Train on less data, by doing for example `clf.fit(X_train.head(7000), y_train.head(7000))`
> 
> I used `%%time` to see how long .fit would take with `max_iter=1000` - it ran for 5 minutes before it finished, but it never converged (aka got an answer)! My computer's old, but if you try that route you should go get a snack.

**Build a logistic classifier and train it on our training data**

In [31]:
clf = LogisticRegression(C=1e9, solver='lbfgs', max_iter=7000)
clf.fit(X_train.head(7000), y_train.head(7000))
clf

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=7000, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

### Examining the coefficients

Use this code to print out the coefficients, it's slightly different than the decision tree version:

```python
import numpy as np

feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients).round(4)
}).sort_values(by='odds ratio', ascending=False)
```

I might throw a `.head(20)` on there so you only see the top ones.

In [33]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients).round(4)
}).sort_values(by='odds ratio', ascending=False).head(20)

feature  coefficient (log odds ratio)    odds ratio
1188  chinese five-spice powder                    134.081670  1.701882e+58
5936       szechwan peppercorns                    104.537241  2.511579e+45
2005                       duck                     89.157240  5.254086e+38
528               black vinegar                     84.546640  5.225635e+36
1542                 corn flour                     79.569654  3.602979e+34
6213                    turnips                     78.969696  1.977439e+34
3585            lo mein noodles                     74.941368  3.520647e+32
3535            light soy sauce                     74.311179  1.874714e+32
1205          chinese rice wine                     72.799568  4.134762e+31
2623       glutinous rice flour                     71.931787  1.736114e+31
5267                   satsumas                     71.253991  8.814854e+30
2279    flowering garlic chives                     68.298373  4.587821e+29
4526                  pineapple                     67.853847  2.941381e+29
2045                egg noodles                     67.789822  2.758962e+29
2986               hoisin sauce                     65.638019  3.207965e+28
6240         unflavored gelatin                     63.888021  5.574617e+27
5348             semolina flour                     61.171447  3.684867e+26
2008               duck breasts                     60.510139  1.902040e+26
1799             dark soy sauce                     60.332579  1.592599e+26
5354                     sesame                     60.240368  1.452310e+26

These are pretty different from our decision tree.

**Why might these numbers be so different compared to our decision tree?** Think about how decision trees and logistic regression work. Or maybe go answer the next question and then come back.

**Use words to describe what the odds ratio for your top ingredient means.**

In [ ]:
# that if the ingredient is Chinese five-spice powder, then the recipe is 134% Chinese (overwhelmingly Chinese)

### Check the confusion matrix

How well does it do compared to the previous two classifiers?

In [35]:
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not chinese', 'chinese'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted not chinese  Predicted chinese
Is not chinese                   9067                202
Is chinese                        182                493

Describe in words how well it does at accurately predicting Chinese food recipes as being Chinese.

### Which was your best classifier so far?

Do you have to think about what "best" means, or is these a clear winner?

# Upgrading to a TF-IDF Vectorizer

Vectorizing using TF-IDF is almost always better than using simple counts. You can't find the most popular words or anything, but common words get downgraded and less common, more descriptive words end up rising in importance. And that seems cool!

Take your `CountVectorizer` code from up above and change it to use a `TfidfVectorizer`. I'll leave the `split_on_commas` code here as a start.

In [44]:
def split_on_commas(text):
    pieces = text.split(",")
    no_spaces = [piece.strip() for piece in pieces if piece]
    return no_spaces

tfidf_vectorizer = TfidfVectorizer(tokenizer=split_on_commas)
X = tfidf_vectorizer.fit_transform(df.ingredient_list)
pd.DataFrame(X.toarray(), columns=tfidf_vectorizer.get_feature_names())

(    oz.) tomato sauce  (   oz.) tomato paste  \
0                         0.0                    0.0   
1                         0.0                    0.0   
2                         0.0                    0.0   
3                         0.0                    0.0   
4                         0.0                    0.0   
5                         0.0                    0.0   
6                         0.0                    0.0   
7                         0.0                    0.0   
8                         0.0                    0.0   
9                         0.0                    0.0   
10                        0.0                    0.0   
11                        0.0                    0.0   
12                        0.0                    0.0   
13                        0.0                    0.0   
14                        0.0                    0.0   
15                        0.0                    0.0   
16                        0.0                    0.0   
17                        0.0                    0.0   
18                        0.0                    0.0   
19                        0.0                    0.0   
20                        0.0                    0.0   
21                        0.0                    0.0   
22                        0.0                    0.0   
23                        0.0                    0.0   
24                        0.0                    0.0   
25                        0.0                    0.0   
26                        0.0                    0.0   
27                        0.0                    0.0   
28                        0.0                    0.0   
29                        0.0                    0.0   
...                       ...                    ...   
39744                     0.0                    0.0   
39745                     0.0                    0.0   
39746                     0.0                    0.0   
39747                     0.0                    0.0   
39748                     0.0                    0.0   
39749                     0.0                    0.0   
39750                     0.0                    0.0   
39751                     0.0                    0.0   
39752                     0.0                    0.0   
39753                     0.0                    0.0   
39754                     0.0                    0.0   
39755                     0.0                    0.0   
39756                     0.0                    0.0   
39757                     0.0                    0.0   
39758                     0.0                    0.0   
39759                     0.0                    0.0   
39760                     0.0                    0.0   
39761                     0.0                    0.0   
39762                     0.0                    0.0   
39763                     0.0                    0.0   
39764                     0.0                    0.0   
39765                     0.0                    0.0   
39766                     0.0                    0.0   
39767                     0.0                    0.0   
39768                     0.0                    0.0   
39769                     0.0                    0.0   
39770                     0.0                    0.0   
39771                     0.0                    0.0   
39772                     0.0                    0.0   
39773                     0.0                    0.0   

       (10 oz.) frozen chopped spinach  (14 oz.) sweetened condensed milk  \
0                                  0.0                                0.0   
1                                  0.0                                0.0   
2                                  0.0                                0.0   
3                                  0.0                                0.0   
4                                  0.0                                0.0   
5                                  0.0                                0.

In [45]:
words3 = pd.DataFrame(X.toarray(), columns=tfidf_vectorizer.get_feature_names())
words3.head()

(    oz.) tomato sauce  (   oz.) tomato paste  \
0                     0.0                    0.0   
1                     0.0                    0.0   
2                     0.0                    0.0   
3                     0.0                    0.0   
4                     0.0                    0.0   

   (10 oz.) frozen chopped spinach  (14 oz.) sweetened condensed milk  \
0                              0.0                                0.0   
1                              0.0                                0.0   
2                              0.0                                0.0   
3                              0.0                                0.0   
4                              0.0                                0.0   

   (14.5 oz.) diced tomatoes  (15 oz.) refried beans  1 inch thick  \
0                        0.0                     0.0           0.0   
1                        0.0                     0.0           0.0   
2                        0.0                     0.0           0.0   
3                        0.0                     0.0           0.0   
4                        0.0                     0.0           0.0   

   1% low-fat buttermilk  1% low-fat chocolate milk  \
0                    0.0                        0.0   
1                    0.0                        0.0   
2                    0.0                        0.0   
3                    0.0                        0.0   
4                    0.0                        0.0   

   1% low-fat cottage cheese  ...  yuzukosho  za'atar  \
0                        0.0  ...        0.0      0.0   
1                        0.0  ...        0.0      0.0   
2                        0.0  ...        0.0      0.0   
3                        0.0  ...        0.0      0.0   
4                        0.0  ...        0.0      0.0   

   zatarains creole seasoning  zatarain’s jambalaya mix  zest  \
0                         0.0                       0.0   0.0   
1                         0.0                       0.0   0.0   
2                         0.0                       0.0   0.0   
3                         0.0                       0.0   0.0   
4                         0.0                       0.0   0.0   

   zesty italian dressing  zinfandel  ziti  zucchini  zucchini blossoms  
0                     0.0        0.0   0.0       0.0                0.0  
1                     0.0        0.0   0.0       0.0                0.0  
2                     0.0        0.0   0.0       0.0                0.0  
3                     0.0        0.0   0.0       0.0                0.0  
4                     0.0        0.0   0.0       0.0                0.0  

[5 rows x 6711 columns]

## Building our features

Create our `X` and `y` variables. Our `X` should be the adjusted word frequencies, and `y` should be whether the recipe is Chinese or not.

In [46]:
X = words3
y = df['is_chinese']

X.shape, y.shape

((39774, 6711), (39774,))

Once that's done, be sure to split your features and labels into test and train datasets.

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Decision trees

**Create a decision tree and train it on our data.** Time how long it takes to complete training.

In [50]:
clf = tree.DecisionTreeClassifier(max_depth=5)
clf = clf.fit(X_train, y_train)
clf

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=5,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

## What does it use to make its decisions?

You can either show me feature importances or the actual tree that's built (if you're using a low enough `max_features`).

In [52]:
feature_names = X.columns
importances = clf.feature_importances_
pd.DataFrame({
   'feature': feature_names,
   'feature importance': importances,
}).sort_values(by='feature importance', ascending=False).head()

feature  feature importance
5667                  soy sauce            0.398109
3535            light soy sauce            0.147308
1552                corn starch            0.121964
3658       low sodium soy sauce            0.094647
1188  chinese five-spice powder            0.056076

Compare this with what you got with the CountVectorizer. Do they seem that different?

### How well does this classifier do?

Confusion matrix, please.

In [53]:
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not chinese', 'chinese'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted not chinese  Predicted chinese
Is not chinese                   9191                 69
Is chinese                        405                279

Does it perform better than the `CountVectorizer` version?

# Logistic Regression classifier

I don't think we did that much better - if any better - with our new version. Let's try one more time using logistic regression.

If you get the "could not converge, increase iterations" error, refer to the tips up above about `max_iter` and `max_features`.

In [54]:
clf = LogisticRegression(C=1e9, solver='lbfgs', max_iter=7000)
clf.fit(X_train.head(7000), y_train.head(7000))
clf

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=7000, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [55]:
feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients).round(4)
}).sort_values(by='odds ratio', ascending=False).head()

feature  coefficient (log odds ratio)     odds ratio
5936       szechwan peppercorns                    232.431603  8.785426e+100
1188  chinese five-spice powder                    196.918859   3.317218e+85
3025              hot chili oil                    158.159553   4.873276e+68
1205          chinese rice wine                    146.871069   6.099506e+63
3535            light soy sauce                    137.163262   3.708924e+59

# Questions

### Question 1: What is "vectorization?" Why do we do it?

In [ ]:
#how many times a word appears in a line or sentence

### Question 2: What's the general difference between a `CountVectorizer` and a `TfidfVectorizer`?

In [ ]:
#TfidVectorizer finds the most important or relevant words

### Question 3: Why did we end up using a custom tokenizer instead of the default scikit-learn one?

In [ ]:
# it made the list of ingredients make sense, so we got "realistic" ingredients

### Question 4: What's the best combination of vectorizer + classifier we've used? Does "best" depend on what's important to us?

In [ ]:
#

### Question 5: How do you feel about trying out random classifiers until you find the "best" one?

### Question 6: What's the most "Chinese" ingredient? You might have a different answer based on what type of classifier you're using - why is that?


### Question 7: If you look at the top few words for the random forest, most of them are various types of soy sauce. How could you improve upon that when if you were spending more time with this dataset?

### Question 8: Scikit-learn vectorizers have a lot of options. Among those are `max_features`, which only takes the top n features, and `min_df`/`max_df`, which only allows words that show up above or below a certain percentage of the documents. For example, `max_df` of 0.75 removes any words that show up in more than 75% of documents. When might these options be useful?


### Question 9: If you had to explain a classifier to someone, would you rather explain a logistic regression or a decision tree? Why is that?

### Question 10: A random forest performs better than a decision tree, but it's easy to see how a decision tree works by looking at the flow chart. If you had to explain to your readers the algorithm behind your analysis, would you choose a better-performing algorithm or one that is easier to explain?

# Bonus!!!

## Multi-class problems

Did you know that `DecisionTreeClassifier` can actually support multiple classes, not just two? Probably not, since I've kept it a secret. If you're interested, let's try this out.

### Convert categories to numbers

First, convert the `cuisine` column into a category with `.astype('category')`

The cuisine looks like a string, but it's now secretly a number. You can use `.cat.codes` to get the code out. Assign that to a new column called `cuisine_code`

**Create a new `X` and `y`**. The `X` is the vectorized words, and the `y` is the cuisine code.

**Test train split!**

**Build a decision tree** and train it on the training data.

The confusion matrix code is a little different, but here you go:

```python
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(df.cuisine.cat.categories)
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)
```